sd/etcd/registrar.go

```diff
@@ -104,17 +105,18 @@ func (r *Registrar) loop() {
 
 // Deregister implements the sd.Registrar interface. Call it when you want your
 // service to be deregistered from etcd, typically just prior to shutdown.
-func (r *Registrar) Deregister() {
-	if err := r.client.Deregister(r.service); err != nil {
+func (r *Registrar) Deregister() error {
+	err := r.client.Deregister(r.service)
+	if err != nil {
 		r.logger.Log("err", err)
 	} else {
 		r.logger.Log("action", "deregister")
 	}
-
 	r.quitmtx.Lock()
 	defer r.quitmtx.Unlock()
 	if r.quit != nil {
 		close(r.quit)
 		r.quit = nil
 	}
+	return err
```

Comment:
Same thing here, if Deregister returns an error, it shouldn't do other stuff.

I think this reveals a fundamental design tension: what should an application do if its service discovery Register or Deregister action fails? Can it do anything useful with that error besides log it? Arguably, Register failure might want to terminate the service, in some circumstances, but if Deregister fails, can/should we do anything?